compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -85,6 +86,64 @@ upgradeTemplates n =
     , "    convert B." <> n <> "{..} = A." <> n <> " {..}"
     ]
 
+-- | Generate the source for a package containing template instances for all templates defined in a
+-- package.
+generateInstancesPkgFromLf ::
+       (LF.PackageRef -> UnitId)
+    -> LF.PackageId
+    -> LF.Package
+    -> [(NormalizedFilePath, String)]
+generateInstancesPkgFromLf getUnitId pkgId pkg =
+    catMaybes
+        [ generateInstanceModule
+            Env
+                { envGetUnitId = getUnitId
+                , envQualify = False
+                , envMod = mod
+                }
+            pkgId
+        | mod <- NM.toList $ LF.packageModules pkg
+        ]
+
+-- | Generate a module containing template/generic instances for all the contained templates.
+generateInstanceModule ::
+       Env -> LF.PackageId -> Maybe (NormalizedFilePath, String)
+generateInstanceModule env externPkgId
+    | not $ null instances =
+        Just
+            ( toNormalizedFilePath modFilePath
+            , unlines $
+              header ++
+              nubSort imports ++
+              map (showSDocForUser fakeDynFlags alwaysQualify . ppr) instances)
+    | otherwise = Nothing
```

Comment:
Is this case just an optimization? I don’t see what goes wrong if we try to generate a module with no instances here. Would be good to document this.

Reply:
it's optimization, I'll add a comment.